src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs

```diff
@@ -104,24 +103,27 @@ public virtual IApplication Start()
                         telemetrySource.WriteTelemetry("Microsoft.AspNet.Hosting.BeginRequest", new { httpContext = httpContext });
                     }
 #pragma warning restore 0618
-                    try
+                    using (logger.RequestScope(httpContext))
                     {
-                        using (logger.IsEnabled(LogLevel.Critical)
-                            ? logger.BeginScope("Request Id: {RequestId}", requestIdentifier) 
-                            : null)
+                        try
                         {
+                            logger.RequestStarting(httpContext);
                             await application(httpContext);
                         }
-                    }
-                    catch (Exception ex)
-                    {
+                        catch (Exception ex)
+                        {
 #pragma warning disable 0618
-                        if (telemetrySource.IsEnabled("Microsoft.AspNet.Hosting.UnhandledException"))
+                            if (telemetrySource.IsEnabled("Microsoft.AspNet.Hosting.UnhandledException"))
+                            {
+                                telemetrySource.WriteTelemetry("Microsoft.AspNet.Hosting.UnhandledException", new { httpContext = httpContext, exception = ex });
+                            }
+#pragma warning restore 0618
+                            throw;
+                        }
+                        finally
                         {
-                            telemetrySource.WriteTelemetry("Microsoft.AspNet.Hosting.UnhandledException", new { httpContext = httpContext, exception = ex });
+                            logger.RequestFinished(httpContext);
```

Comment:
It's odd to log the response when there's an unhandled exception. E.g. the status code will probably be 200 until after the server catches and handles the exception. Maybe FinishRequest(exception) for the failure case?

The response also may not be complete at this point. E.g. if there is no body, or if the body was buffered, OnStarting may not have fired yet. You may be logging incomplete headers.